### ISSUE TITLE
Computation graphs and neural networks

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#1995]
Quoting @jchodera's feature request at https://github.com/pandegroup/openmm/issues/1980#issuecomment-362759139:

> In order to facilitate interoperability with machine learning, or even just simplify the implementation of complex forces or CVs, the ability to create a Force from a computation graph (such as a TensorFlow tf.Graph or pytorch compute graph.
>
>It may be difficult to support all Ops, but supporting a basic subset may still be useful.

I'd like to start discussion going on what this means and how to implement it.  And especially what it does *not* mean, since the description above is very open ended and we need to narrow it down to create something practical.  So here is a first attempt at a set of requirements.  This is just intended as a starting point for discussion.  Feel free to suggest different requirements.

1. The goal of this feature is to allow forces to be defined by neural networks.  You provide a network that takes atom coordinates as input and produces an energy as output.

2. This needs to support arbitrary networks.  That includes complex network topologies, convolutional layers, etc.  Just about anything you can define in Tensorflow.  A simplified mechanism that only supported MLPs, for example, would not be sufficient.  (We might omit support for recurrent networks, though, if that turns out to be difficult.)

3. We will *not* attempt to use networks in ways more complicated than described in 1.  There are all sorts of other things we could imagine doing: built in support for neighbor lists to optimize atomic convolutions, using a neural network to post-process the results of other forces, etc.  You can implement anything that can be described with a compute graph, but it's up to you to implement it yourself.

4. We will only support inference, not training.  You define all the network weights before creating the context, and then they don't change.

5. We will make reasonable efforts to provide good performance, but this feature prioritizes flexibility over speed.  For example, the lack of neighbor list support mentioned in 3.  Think of this as primarily a research tool.  Once a method is fully developed, you might decide to write a plugin to implement the calculation is a more optimal way.

**Comment by jchodera** [comment]
Thanks for getting this started!

Some comments:
* We would like to allow the use of the compute graph to either compute energies (whose autodiff-computed forces would be used in computing forces for dynamics) or the values of collective variables
* Compute graphs are more general than "networks". It's simply a way to specify a directed acyclic graph of operations (Ops), which may encode a feed-forward network of some sort, or may not.
* While supporting the particle positions as input features is simplest, I think it's an open question whether other featurizations may be of interest. What might people want beyond the atomic coordinates?
* Supporting the use of a fixed compute graph sounds totally fine; training can always be done outside OpenMM.
* Prioritizing flexibility over speed is reasonable, but the primary goal of implementing this feature within OpenMM (instead of calling out to TensorFlow or PyTorch) would be to avoid the overhead of pulling coordinates every timestep when using this feature to compute forces or CVs, so it's not quite true that performance is entirely irrelevant.

**Comment by peastman** [comment]
> We would like to allow the use of the compute graph to either compute energies (whose autodiff-computed forces would be used in computing forces for dynamics) or the values of collective variables

That shouldn't be a problem.  If it's a Force, you can stick it into a CustomCVForce.

> What might people want beyond the atomic coordinates?

Anything they can compute from the atomic coordinates using a compute graph? :)

> so it's not quite true that performance is entirely irrelevant.

Agreed.  We'll certainly make reasonable efforts to enable good performance.  But compute graphs are inherently limited when it comes to performance.  You can describe almost any calculation, but it won't always be done in the most efficient way.

**Comment by jchodera** [comment]
> Anything they can compute from the atomic coordinates using a compute graph? :)

In the spirit of building an extensible API, what about allowing for the possibility of additional featurizations in the future, but starting with `CartesianFeaturizer` (perhaps allowing a subset of atoms to be specified) as the only implemented featurization scheme to begin with?

I could imagine others wanting to use something like atom-centric K-nearest-neighbor feature spaces in the future, but given that this is still such an area of active research, it may be best to wait a bit to see how things shake out before implementing more featurization strategies.

**Comment by peastman** [comment]
That's the sort of complication I was trying to avoid with item 3.  It adds a lot of complexity to the architecture, but doesn't actually make it much more powerful.  Just about any featurization you want can be implemented with a compute graph.  Why add a new class of plugins to enable computations that are already possible (and likely easier to implement) with the mechanism we just added?

**Comment by jchodera** [comment]
I'm thinking there may be interest in implementing things like ANI-1 [http://doi.org/0.1039/C6SC05720A] which uses a local neighborhood around each atom to compute energies. While this could be implemented as a compute graph that computes the local features for each atom, it's hard for me to grasp how this could be efficient without some sort of neighborlist.

**Comment by samarjeet** [comment]
Ah, looks like we are working on very similar thing here. We have a
working-ish system wherein we have been creating the Force (NNQMForce -
which inherits from Force). Feature vector that we learn is similar to ANI
mentioned earlier however it also includes identity of the atom as well (at
the cost of sparsity). We tried using Tensorflow but its C++ API was pretty
undocumented, so  currently we are training with Caffe2. Although we are
planning to do the training in OpenMM itself but the user would be expected
to just import the model just like any other normal force field. CC'ing
Jing Huang in this one.

**Comment by peastman** [comment]
> it's hard for me to grasp how this could be efficient without some sort of neighborlist.

We've got a pure Tensorflow implementation of neighbor lists in DeepChem.  It uses a basic cell list algorithm, so not as fancy as what's in OpenMM, but it does the job.  We also have an implementation of ANI-1.

@samarjeet Thanks for the suggestion!  I haven't looked at Caffe at all, but clearly I should.  Does it have a reasonably straightforward way a user could define a computation graph and import it into the OpenMM force?

**Comment by msultan** [comment]
This feature would be amazing to have. I have done some testing (#1973 and internally) and found that the CustomCVForce can be wrangled to accept rather large and complex compute graph. The only downside is that it requires knowing how each layer/op works and being able to express it as a string. 

If we had to pick an API, I would like to **heavily** lobby for PyTorch because of its pythonic interface and dynamic graph construction.

**Comment by jchodera** [comment]
@msultan : I bet it's possible to support multiple compute graph formats, at least from the Python `app` layer since it's so easy to traverse these compute graphs in Python in most of these tools.

**Comment by peastman** [comment]
Tensorflow, PyTorch, and Caffe2 have been suggested as possible libraries to use for doing the computation.  Here are the main requirements I see for a computation library.

- It needs to have a stable, sufficiently powerful C++ API to do everything we need.  All the low level code will be written in C++.
- It needs to be possible to use in a pure C++ program.  We may want to use it in contexts where Python isn't available, such as on Folding@Home.  It's OK if you use Python to define the graph, as long as that graph can then be saved to a file that can be loaded in using only C++.
- If possible, it would be good if it provides access to data on the GPU.  We'll be copying positions and forces back and forth, and it will help performance if we can do that directly on the GPU without bringing it back to the CPU.

Does anyone have experience with how well the different libraries meet these requirements?

**Comment by peastman** [comment]
Answering part of my own question.  It looks like PyTorch's C++ support is very limited and not mature.  However, they do make it easy to export models in a format that can be loaded with Caffe2: http://pytorch.org/docs/master/onnx.html.

**Comment by jchodera** [comment]
@peastman: Do you want to offload these op graph computations to an external library, or would you rather implement a minimal GPU-based graph computation engine that would be unique to OpenMM (much like the Lepton-based engines are)?

If an external library, would you want something that is ingested into OpenMM, something that OpenMM uses as an external dependency, or something that can be installed optionally and used if the user has installed it?

I imagine that there are advantages to sticking with the standard OpenMM approach where a minimal, simple, extensible C++ API exposes general functionality, a Python wrapper is created, and the "app" layer provides a more fully-featured interface for automatic conversion from other file types.

There are some small C++ CUDA-enabled op graph computation libraries floating around that may be worth looking at, like [this one](https://github.com/joschu/cgt?files=1).

**Comment by peastman** [comment]
I don't see much advantage in reimplementing this ourselves.  We want full support for arbitrary computation graphs with all the features you expect in a full featured deep learning library.  And it needs to run efficiently on both CPUs and GPUs.  And we want to be able to construct computation graphs using the APIs of popular toolkits.  We're not talking about a minimalist computation engine.  In fact, this is pretty much a perfect description of something like Caffe2.

My plan is that at least initially, this will be a plugin that's installed separately, similar to the Plumed plugin.  We aren't going to bundle a large deep learning library with OpenMM itself, nor make one into a dependency for it.

**Comment by jchodera** [comment]
> My plan is that at least initially, this will be a plugin that's installed separately, similar to the Plumed plugin. We aren't going to bundle a large deep learning library with OpenMM itself, nor make one into a dependency for it.

Sounds good. We can easily make that conda-installable, and have it pull in the necessary dependencies since [caffe2 seems to be conda-installable](https://anaconda.org/caffe2/caffe2).

We should add one more use case to the list: Integrators! @pgrinaway has been asking for support for more complex generalized integrators for some time, and I've just found out that Ben Leimkuhler is working on a suite of high-quality integrators for TensorFlow.

**Comment by jchodera** [comment]
Hm, PyTorch has recently added support for "tensor comprehension", which sounds similar to Lepton for tensors: http://pytorch.org/2018/03/05/tensor-comprehensions.html

**Comment by peastman** [comment]
Very cool!

**Comment by peastman** [comment]
I have a *very* preliminary version of this written!  Don't expect too much from it yet in terms of speed, features, or lack of bugs, but it is functional.  https://github.com/pandegroup/openmm-nn

First install Caffe2 by following the instructions at https://caffe2.ai/docs/getting-started.html.  Then compile and install the plugin in the usual way.

To use it, you create a model and save it to disk.  Caffe2 models are saved as two files, one containing the "init graph" and the other containing the "predict graph".  If you create your model using Caffe2's Python API, you can save it like this:

```python
with open('predict_net.pb', 'wb') as f:
    f.write(m.net._net.SerializeToString())
with open('init_net.pb', 'wb') as f:
    f.write(m.param_init_net._net.SerializeToString())
```

where `m` is a `ModelHelper`.  If you want to create it with PyTorch (or [various other tools](https://github.com/onnx/tutorials)), you can do that by [exporting to ONNX format](https://github.com/onnx/tutorials/blob/master/tutorials/PytorchOnnxExport.ipynb) and then [converting that](https://github.com/onnx/tutorials/blob/master/tutorials/OnnxCaffe2Import.ipynb) to Caffe2 files.

The model should have a single input called `positions`.  It should be a tensor of shape (# atoms, 3).  It should produce two outputs: a scalar called `energy` and a tensor of shape (# atoms, 3) called `forces`.

Now you can create a force like this:

```python
from openmmnn import *
f = NeuralNetworkForce("predict_net.pb", "init_net.pb")
system.addForce(f)
```

**Comment by jchodera** [comment]
Awesome!

Since [ONNX](https://github.com/onnx/onnx) seems to be the general open exchange format for neural networks, what if we add a Python-level helper force so you can simply do:
```
import onnx
from openmmnn import *
onnx_model = onnx.load('awesomenet.onnx')
f = ONNXForce(onnx_model)
system.addForce(f)
```
This would make it easier for everyone to [export their networks from any tool](https://github.com/onnx/tutorials) and easily have the OpenMM plugin (which could have `onnx` as a dependency) be able to use them directly.

**Comment by peastman** [comment]
Definitely.  It would be even better if the C++ code could support that directly, but all the ONNX parsing seems to be in Python.

**Comment by jchodera** [comment]
@peastman: Do you intend the plugin to be usable via the C++ API as well, or just via Python?

I'm thinking about how we might want to package and distribute the plugin to make it easy for people to use, as well as make sure we can bring in the appropriate dependencies (caffe2 and onnx) automatically. Is this currently intended to be built and packaged like OpenMM?

**Comment by peastman** [comment]
It definitely needs to work from C++.  We may want to use it in Folding@Home at some point.

Ultimately there should be a conda package for it.  Unfortunately, we won't be able to directly declare Caffe2 as a dependency, because they have different packages depending on whether it's using a GPU and what version of CUDA you have installed.  We'll have to tell people to separately install the correct Caffe2 package first.

**Comment by jchodera** [comment]
We should probably move to a model that is similar to what pytorch uses:
https://github.com/pytorch/builder/tree/master/conda

They have packages like `cuda90` and `cuda91` that provide a "feature" with the same name. The features you have installed controls which version of the pytorch package you get when you `conda install pytorch`. I've been investigating this mechanism for OpenMM as well. 

We'd then have a recommended scheme like:
```bash
# Install the cuda91 feature to indicate you have CUDA 9.1 installed
conda install cuda91
# Install the appropriate version of OpenMM and the plugin
conda install openmm openmmnn
```

Tagging @Lnaden, who will be interested in checking this out as well.

**Comment by peastman** [comment]
Caffe2 development was just merged into the PyTorch repository, so *maybe* they'll eventually move to that model too.  Currently, installing Caffe2 for GPU is a fairly manual process, including downloading various packages from Nvidia's website and installing them by hand.  In fact, I haven't gotten it to work yet, though that's probably just because they don't have a package for CUDA 9.1.

**Comment by jchodera** [comment]
That's great to hear!

@peastman : What do you think about also extending the `CustomIntegrator` API to support steps that involve compute graphs acting on positions, velocities, forces, energies, and/or global and local variables? Is that something that could be performant? This would be a really neat but simple way to encode much more complex intermediate steps.

**Comment by peastman** [comment]
CustomIntegrator is part of the core library.  So we would need to either completely reproduce it in the plugin, or else make Caffe2 into a dependency for the core library.  Neither of those is too appealing.  Also, Caffe2 doesn't support double precision, which is OK for computing forces but probably not good enough for integration.

**Comment by jchodera** [comment]
Could the `CustomComputeGraphIntegrator` be a *subclass* of `CustomIntegrator`, adding methods like
```python
integrator.addComputeGraphStep(onnx_model, global_variables=['global1', 'global2'], dof_variables=['dof1', 'dof2'])
```
such that the list of global and per-dof variables expected as input and output for the model are specified? That may not even be necessary if there is a way to detect what input/output names are specified in the model.

**Comment by jchodera** [comment]
One other observation: In the [`NeuralNetworkForce` API](https://github.com/pandegroup/openmm-nn/blob/master/openmmapi/include/NeuralNetworkForce.h#L46-L55), do we really want to specify the caffe2 models as strings, rather than `NetDef` objects?

**Comment by peastman** [comment]
> do we really want to specify the caffe2 models as strings, rather than NetDef objects?

Everything in the current implementation is *very* preliminary.  I'm not even 100% certain I'll stick with Caffe2 rather than switching to something else.

**Comment by peastman** [comment]
I gave up on waiting for Caffe2 and PyTorch to mature enough to be usable for this, and instead converted it to use TensorFlow.  I think that's going to be a much better base for it.

To use it, install the TensorFlow C API as described at https://www.tensorflow.org/install/install_c.  You should then be able to build the plugin.  If you installed the API somewhere other than /usr/local, set the CMake variable `TENSORFLOW_DIR` to point to the directory where it's installed.

Create your model with Python and save it to a binary protocol buffer file.  It should have an input called "positions" of shape (n_atoms, 3), a scalar output called "energy", and an output of shape (n_atoms, 3) called "forces".  Here's a very simple example:

```python
import tensorflow as tf
graph = tf.Graph()
with graph.as_default():
    positions = tf.placeholder(tf.float32, [None, 3], 'positions')
    energy = tf.reduce_sum(positions**2, name='energy')
    forces = tf.identity(tf.gradients(-energy, positions), name='forces')
tf.train.write_graph(graph, '.', 'graph.pb', as_text=False)
```

I also added support for periodic boundary conditions.  To use them you need to do two things.  First, add an input of shape (3, 3) called "boxvectors" to the model.  Second, call `setUsesPeriodicBoundaryConditions(True)` on your NeuralNetworkForce object.  You can then use the box vectors as part of your graph.  For example,

```python
positions = tf.placeholder(tf.float32, [None, 3], 'positions')
boxvectors = tf.placeholder(tf.float32, [3, 3], 'boxvectors')
boxsize = tf.diag_part(boxvectors)
periodicPositions = positions - tf.floor(positions/boxsize)*boxsize
```

If you've trained a model that involves variables, you can use TensorFlow's `freeze_graph` script as decribed at https://www.tensorflow.org/extend/tool_developers/#freezing.  That creates a version of the model in which all variables are turned into constants with their final trained values.

**Comment by peastman** [comment]
> can we spruce up your compute graph plugin repo to something where we can start building conda-installable packages that also install TensorFlow?

If you can figure out a way to do it, that would be great.  The problem is that it requires the TensorFlow C API, which has to be downloaded separately from https://www.tensorflow.org/install/lang_c.  It isn't included with the conda package.

> For example, there are other features that OpenMM could efficiently compute that might be more suitable than Cartesian coordinates for defining interaction potentials.

Open issues on the openmm-nn repo for any features you'd like to see added.

**Comment by jchodera** [comment]
We just came across this TensorFlow plugin for HOOMD:
https://github.com/ur-whitelab/hoomd-tf
Seems to have some neat ideas!

**Comment by DanielWicz** [comment]
Generally using TF1.0 in @peastman 's plugin is a big problem, because it is obsolete now and TF1.0 is really hard to use in comparison to TF2.0.

**Comment by peastman** [comment]
It's on my list to fix.  TensorFlow 2.0 dropped support for the C API, so supporting it was impossible.  2.3 finally added it back again.

**Comment by peastman** [comment]
Using TensorFlow 2.3.1, the existing code compiles without problem.  The reference and OpenCL tests also run perfectly, but the CUDA tests tend to segfault deep inside TensorFlow code.  I'm trying to figure out why.  The code for the OpenCL and CUDA implementations are very similar.

**Comment by jchodera** [comment]
We should stand up nightly CI for these plugin repos as soon as we can!

**Comment by DanielWicz** [comment]
>  segfault
Which version of CUDA, cudann and Python do you have ? Generally Tensorflow is very picky with regard to the combination of CUDA, cudann and Python. Some of them just don't work.

**Comment by swails** [comment]
> We should stand up nightly CI for these plugin repos as soon as we can!

If you add the necessary pre-reqs to https://github.com/Amber-MD/common-dockerfiles/blob/master/cuda-opencl/Dockerfile, Jenkins can start testing it anytime.

**Comment by peastman** [comment]
This is a C++ program, so there's no Python involved.  It's possible this is a version inconsistency.  I have CUDA 11.1 in my `LD_LIBRARY_PATH`, but when I run it I see TensorFlow printing out messages like, "Successfully opened dynamic library libcudart.so.10.1".  Any idea how it figures out what version to load?

**Comment by DanielWicz** [comment]
> This is a C++ program, so there's no Python involved. It's possible this is a version inconsistency. I have CUDA 11.1 in my `LD_LIBRARY_PATH`, but when I run it I see TensorFlow printing out messages like, "Successfully opened dynamic library libcudart.so.10.1". Any idea how it figures out what version to load?

Check `CUDA_HOME` and `DYLD_LIBRARY_PATH`, the other may be from conda env. Actually two different versions of CUDA may be the case, why you get segfaults while compiling Tensorflow's CUDA part.
